### PR TITLE
Poetry2reqs

### LINKED ISSUE #227: `micropipenv requirements --method poetry --no-default` won't generate sub-dependencies if they're also a main dependency
**Describe the bug**

`micropipenv requirements --method poetry --no-default` won't generate sub-dependencies if they're also a main dependency.

e.g. if I declare `python-dateutil` to be a project dependency, and `freezegun` to be a dev dependency, and I run `micropipenv requirements --method poetry --no-default > dev-requirements.txt` and then try to pip install that, I will get the following error:

```
ERROR: In --require-hashes mode, all requirements must have their versions pinned with ==. These do not:
    python-dateutil>=2.7 from https://files.pythonhosted.org/packages/36/7a/87837f39d0296e723bb9b62bbb257d0355c7f6128853c78955f57342a56d/python_dateutil-2.8.2-py2.py3-none-any.whl#sha256=961d03dc3453ebbc59dbdea9e4e11c5651520a876d0f4db161e8674aae935da9 (from freezegun==1.1.0->-r dev-requirements.txt (line 13))
```
because freezegun requires python-dateutil, but it wasn't added to the requirements.txt file

**To Reproduce**

pyproject.toml
```
[tool.poetry]
name = "test"
version = "1.0.0"
description = ""
authors = ["Your Name <you@example.com>"]
readme = "README.md"

[tool.poetry.dependencies]
python = "^3.8"
python-dateutil = "2.8.1"

[tool.poetry.dev-dependencies]
freezegun = "1.1.0"

[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"
```

1. micropipenv requirements --method poetry --no-default > requirements.txt
2. pip install -r requirements.txt
3. Error

**Expected behavior**
Ideally, if a main dependency is also a transitive dependency of a dev dependency, that dependency would be added to the output of --no-default

### PR DESCRIPTION
**Pull request by frenzymadness**
## Related Issues and Dependencies
Fixes #227 
…

## This introduces a breaking change

- [ ] Yes
- [x] No

(According to the existing tests)

During implementing the fix, I've realized that the process of transferring poetry config to requirements file is not ideal and it also seemed not tested well.

Even all the tests are passing, we should review this carefully. I'm still not sure I understand all the bells and whistles in micropipenv.

### CODE REVIEW

**Review by fridex — COMMENTED** [review]
Thanks for the fix. Given the approach that poetry has taken (features outside of the Python packaging standards), it might be really hard to fully support poetry.

File: micropipenv.py [hunk 1]
```diff
@@ -723,8 +724,13 @@ def _poetry2pipfile_lock(
 
     additional_markers = defaultdict(list)
     skip_all_markers = object()
+    add_to_dev = list()
 
     for entry in poetry_lock["package"]:
+
+        if entry["category"] not in ("dev", "main"):
+            raise PoetryError("Unknown category for package {}: {}".format(entry["name"], entry["category"]))
```
[the remark this answers is not in the document]
Comment (frenzymadness) [reply micropipenv.py:732]: Sure, why not. The error is just moved from the bottom of the function.

[the remark this answers is not in the document]
Comment (frenzymadness) [reply micropipenv.py]: Fixed, added test, rebased.

**Comment by frenzymadness** [comment]
Well, if this change is okay we can merge it, and then, if future requests to support some special functionalities from poetry appear, we can evaluate them case by case.

**Comment by fridex** [comment]
> Well, if this change is okay we can merge it, and then, if future requests to support some special functionalities from poetry appear, we can evaluate them case by case.

Sounds reasonable 👍🏻

**Review by fridex — APPROVED** [review]
looks good to me, thanks! 💯

**Pull request merged**